### DISCUSSION TITLE
unusual response time patterns and IOSQE_ASYNC

### DISCUSSION THREAD

**Discussion by carns** [axboe/liburing#847] [Q&A]
I've been running some experiments to see how to best extract performance from the local NVMe drives on the Frontier system at ORNL (https://www.olcf.ornl.gov/frontier/ and https://docs.olcf.ornl.gov/systems/frontier_user_guide.html).  Each compute node has two high-end Samsung NVMe drives in a RAID-0 configuration formatted with XFS.  The kernel version is 5.14.21 (and is unlikely to be significantly upgraded any time soon).

I'll focus on one specific workload to narrow down the scope of my question.  This performance graph shows the response time of 256 KiB pread operations issued using 16 pthreads, 16 distinct files, and O_DIRECT (no liburing, this is just for a baseline):

![posix-read-directsync-256k-16x frontier latencies](https://user-images.githubusercontent.com/5541937/233170042-65e3a757-d6d3-46e9-8d3e-4f11f3b7d7c5.png)

If I mimic the same thing using liburing (with 1 thread set aside to harvest completion queue events) then I get this odd graph:

![liburing-no-async-read-directsync-256k-16x frontier latencies](https://user-images.githubusercontent.com/5541937/233170799-2ff4dc32-3e96-4ced-bfd4-d6c3edf61544.png)

The response time values are "floating" as if there is a fixed cost of ~180 usec imposed on almost every operation.  This is very reproducible, and I'm only getting about a third of the expected bandwidth (the first graph was from a run that approached ~60 GiB/s, while this one was more like ~20 GiB/s).

If I add `IOSQE_ASYNC` to the submission, however, performance rebounds and roughly matches what I was getting with pthreads issuing preads.  If I plot the response time of individual operations then the "floating" phenomenon is gone.

![liburing-async-read-directsync-256k-16x frontier latencies](https://user-images.githubusercontent.com/5541937/233171592-baf85d1f-baef-4b1b-bf11-9864f0edde77.png)

Superficially, it looks like I need to set that flag to get good uring performance on this system.  Thinking deeper, though, this looks less like a performance optimization and more like a workaround for a bug.  The response times in that 2nd plot don't make any sense.

Does anyone have any insight on why I might be seeing this behavior or why the IOSQE_ASYNC flag has such a dramatic effect?

For some broader context: I believe that I am seeing the same thing on a different system (Polaris at ANL: https://www.alcf.anl.gov/polaris) which also has dual NVMe drives in a RAID 0 configuration, but with EXT4 instead of XFS.  Unfortunately the kernel on that machine is 5.3.18 and does not support IOSQE_ASYNC, so I cannot test the "fix" there, I just see the weird response times.  It will eventually be upgraded, but not in the immediate term.

I've also tested this same workload on my laptop (with a good single consumer NVMe circa 2019 and a very recent kernel) and do not see this behavior at all.  In fact the IOSQE_ASYNC flag sometimes hurts performance, as cautioned by the liburing man pages :)  I don't know if the discrepancy is caused by the hardware differences or the kernel version.

**Comment by isilence** [comment]
How many threads there was for io_uring?

The first guess is that it's single threaded and CPU bound, whereas `IOSQE_ASYNC` creates worker threads effectively increasing the number of CPUs it's executed on. We'd need to know the utilisation of the CPU / core it was running on.

**Reply by carns** [reply]
Ah, Ok.  I am doing all of the `io_uring_submit()` calls from a single thread/core.  There is one other thread/core that is doing all of the `io_uring_wait_cqe()` calls.

So it is likely that I am overloading the thread in charge of submissions when I'm not using  `IOSQE_ASYNC`?  It did not occur to me that this would be the case, but I have not looked at what happens in the io_uring code path after submission.

I'll see if I can profile the thread-level CPU usage.  It is also plausible for me to issue operations from multiple threads, but I was hoping to avoid that for my use case.

**Reply by isilence** [reply]
Yes, there is a good chance for that. Usually most of overhead is in lower layers like XFS and kernel's block layer, even if it's an async submission, and io_uring cannot do much about it.

> I'll see if I can profile the thread-level CPU usage. It is also plausible for me to issue operations from multiple threads, but I was hoping to avoid that for my use case.

Just pin the task with `taskset` and run `top` / `mpstat` / `sar` / etc. looking for util of that CPU you pinned io_uring to, should be enough to confirm.

**Reply by isilence** [reply]
@carns, how did it go? Did I guess right or is it something else?

**Reply by carns** [reply]
@isilence Unfortunately I've had a string of travel and some other commitments and haven't circled back to it just yet.  I will definitely share results when I do.

**Comment by bsbernd** [comment]
I would use 'htop' instead of top, to see if the single thread is bouncing between cpus. That might as well kill your performance. Also check how much cpu time your your io_uring_wait_cqe() thread actually needs and it shouldn't be better run on the same cpu as the io_uring_submit() thread. I don't know if it also applies to your workload, but with fuse-over-uring io submission and io processing thread should run on the same core for sync requests, for single threaded application it has a gain of factor 4, for multi threaded there is no scaling without it.

**Reply by isilence** [reply]
Forgot to mention above, but the idiomatic way is to have a ring per thread, which will do both submissions and waiting for completions, then it can be paired with `IORING_SETUP_DEFER_TASKRUN` and other optimisations. The communication between rings can be done via `IORING_OP_MSG_RING`. Just a general advice, it might be not a fit for fuse.

**Reply by carns** [reply]
I have a quick question about this idiom (in which there is one ring per thread, with that thread doing both submissions and completions).  Assuming the actual I/O requests are originating elsewhere (in my case, remote clients of a distributed service), then I need a way to break a thread out of its completion wait so that it can generate a new entry for the submission queue in a timely manner.  Is that what you are getting at with the IORING_OP_MSG_RING mechanism?  If so, is the expectation that waking a ring's thread so that it can inject the I/O operation on the caller's behalf is faster than the caller injecting I/O operations into the ring directly?

Hopefully that question makes sense.  I think this may have been a fundamental misunderstanding on my part.

**Reply by isilence** [reply]
The original use case for `IORING_OP_MSG_RING`, I believe, is when threads themself initiate and drive IO but need eventual synchronsation with other threads. A good example would accepting a socket and then handing it off to another thread, so load balancing the work.

Sharing submission queue b/w threads is usually a bad idea as it would require locking around submission. The alternatives are either to use `IORING_OP_MSG_RING` or do it via some file like eventfd (e.g. write to evenfd on submission, the target ring keeps one inflight read of the eventfd and rearms it as soon as it gets an event). Depending on the rate of such submissions and the use case, it might be better for `IORING_OP_MSG_RING` cqe to carry more than one IO, perhaps with extra atomics. I guess it needs to be tried how well it performs  for your case.

**Comment by bsbernd** [comment]
Yeah, the fuse uring patches have a thread per cpu and each thread has its own ring (except in debug/devel mode, which has just one ring). Interesting part is that all these extra flags make it a bit slower - not much, but noticeable.  Will investigate later, when I'm through with more critical issues (scheduler for now).

**Comment by carns** [comment]
Just to close out this discussion; I went back and re-architected the code that I had so that there is exactly one thread in charge of submission and completion on a given ring.  The number of rings in use is configurable. External callers (which may come from an arbitrary number of external threads; I can't control that part) enqueue operations into linked lists in user space and then wake up one of the ring threads to pull it off of the queue and handle liburing submission and completion.  This seems to work fine and makes life easier for liburing.  The rate particularly for small operations is great despite the extra steps needed in my use case to avoid uring contention.

FWIW I'm waking uring workers that are blocked waiting for completion by using an eventfd that is being watched with a multishot poll so that I can get some io depth going in each ring when there is a lot of large I/O activity.  This also seems to work Ok.  (I realized that the OP_MSG thing is more for one uring thread to signal another uring thread, not for an external thread to signal a uring thread).  I can do some more profiling/tuning here to get the frequency of signalling right for different situations, but the basic method is fine.

I still need to implement an alternative path for platforms that I need to use that don't have multishot poll, but I'll work that out.  Doesn't seem like it will be a big deal.

Thanks for the comments all!

**Comment by carns** [comment]
On my laptop "big" reads (256 KiB) still seem slower with uring than with a pool of threads blocking on plain old pread() calls, but I'm not going to read too much into that until I can benchmark some more platforms.  I can't imagine why this is unless it's just a memory locality thing or something.
